src/full/Agda/Compiler/Malfunction/AST.hs

```diff
@@ -0,0 +1,290 @@
+{- |
+Module      :  Agda.Compiler.Malfunction.AST
+Maintainer  :  janmasrovira@gmail.com, hanghj@student.chalmers.se
+
+This module defines the abstract syntax of
+<https://github.com/stedolan/malfunction Malfunction>. Please see the
+<https://github.com/stedolan/malfunction/blob/master/docs/spec.md Malfunction
+language specification>
+-}
+{-# LANGUAGE OverloadedStrings #-}
+module Agda.Compiler.Malfunction.AST
+  ( IntType(..)
+  , IntConst(..)
+  , UnaryIntOp(..)
+  , BinaryIntOp(..)
+  , VectorType(..)
+  , Mutability(..)
+  , BlockTag
+  , Case(..)
+  , Ident
+  , Longident
+  , Mod(..)
+  , Term(..)
+  , Binding(..)
+  -- NOTE: I don't know which of these is preferable
+  --  * Don't re-export anything from Agda.Utils.Pretty
+  --  * export a few things (like we do currently)
+  --  * Re-export the whole module
+  , pretty
+  , prettyShow
+  ) where
+
+import Data.Int
+-- There does exist a definition of a type-class `Pretty` in the package
+-- `pretty` but this is not the one used for Treeless terms, so for consistency,
+-- let's go with Agda's choice.
+import Agda.Utils.Pretty
+
+-- | The integer types.
+data IntType
+  = TInt
+  | TInt32
+  | TInt64
+  | TBigint
+  deriving (Show, Eq)
+
+-- | An integer value tagged with its corresponding type.
+data IntConst
+  -- Int may be the wrong type.
+  --
+  -- In malfunction Int is:
+  --
+  -- > `int` uses either 31- or 63- bit two's complement arithmetic
+  -- > (depending on system word size, and also wrapping on overflow)
+  -- > https://github.com/stedolan/malfunction/blob/master/docs/spec.md
+  --
+  -- And in Haskell:
+  -- > A fixed-precision integer type with at least the range
+  -- > [-2^29 .. 2^29-1]
+  -- > https://hackage.haskell.org/package/base-4.9.0.0/docs/Data-Int.html
+  --
+  -- Jan: Just run
+  -- logBase 2 ((fromIntegral (maxBound :: Int)) :: Double)
+  -- in my computer (64 bits) --> 2 ^ 63 - 1 == (maxBound :: Int)
```

Comment:
Agda officially (builtin/stdlib) only has arbitrary-precision integers, although one can add ``Int32`` and co. using the FFI. If operations on those types should reduce during type checking, we would need to precisely specify their range and overflow behaviour. See e.g. https://github.com/agda/agda/issues/2169 , which shows what happens if we are imprecise.